Allow typed dictionaries, and dictionary subscripts from expander

douglasjacobsen opened the pull request:
This merge enables the expander to handle dictionaries better.

The most obvious change from this is that given:
```
variables:
  test_dict:
    key1: val1
    key2: val2
```
then `'{test_dict["key1"]}'` should expand to `"val1"`.

A more subtle change from this is that given the variable definition above, inside object definitions one can now do:
```
test_dict = app_inst.expander.expand_var_name("test_dict", typed=True)

for key, val in test_dict.items():
  print(f"Key {key} = {val}")
```

Tests and documentation have also been added for this.

douglasjacobsen committed:
Allow expander to return typed dictionaries

This commit adds the ability for the expander to return a typed
dictionary based on variable definitions.

Subscripting of dictionaries will also return the value of the key, from
the expander now.
i.e. `'{dict["key"]}'` will return the value of the key.

douglasjacobsen committed:
Update tests and documentation

linsword13 commented on lib/ramble/ramble/expander.py:
```diff
@@ -906,24 +906,41 @@ def _eval_unary_ops(self, node):
+                    # DEPRECATED: ast.Index was dropped in python 3.9
+                    if hasattr(ast, "Index") and isinstance(slice_node, ast.Index):
+                        key = self.eval_math(slice_node.value)
+                        return op_dict[key]
```
What happens when key is not in the dict?

douglasjacobsen replied:
Ah good catch. Currently, you get a `KeyError`, but I don't think that's what we want.

I'll work on fixing this.

douglasjacobsen replied:
Fixed!

douglasjacobsen committed:
Update error handling in expander ast logic

This commit updates the error handling in the expander ast logic so that
SyntaxErrors trigger a debug print before being raised again (and
causing an unexpanded string to be returned)

douglasjacobsen committed:
Unify SyntaxError debugging information for expander

This commit unifies the handling of generic SyntaxErrors in Ramble's
expander. This allows all SyntaxErrors to print debugging information,
which can be useful in determining why the expander behaves the way it
does.
